### PR TITLE
add ability to translate the user guide on Crowdin

### PR DESCRIPTION
**Pull request by michaelDCurran**
### Link to issue number:
None.

### Summary of the issue:
Currently user documentation such as the user guide are made translatable via a custom (and very old) translation system hosted by NV Access. For many reasons we need to move away from this old system to something more mainstream and  maintainable. We have already successfully moved translation of NVDA interface messages to Crowdin, and we should do the same for the user guide and other documentation.
 
### Description of user facing changes
None.

### Description of development approach
* Added markdownTranslate.py, which contains several commands for generating and updating  xliff files from markdown files. These xliff files can then be uploaded to Crowdin for translation, and eventually downloaded again and converted back to markdown files. Commands include:
    * generateXliff: to generate an xliff file from a markdown file. Firstly a 'skeleton' of the markdown file is produced which is all the structure of a markdown file, but the translatable content on each line has been replaced by a special translation ID. Lines such as blank lines, hidden header rows, or table header separator lines are included in the skeleton in tact and are not available for translation. The xliff file is then produced, which contains one translatable string per translation unit, keyed by its respective translation ID. Each unit also contains translator notes to aide in translation, such as the line number, and any prefix or suffix markdown structure. E.g. a heading might have a prefix of `###` and a suffix of `{#Intro}`. The skeleton is also embedded into the xliff file so that it is possible to update the xliff file keeping existing translation IDs, and or generate the existing markdown file from the xliff file.
    * generateMarkdown: Given an xliff file, a the original markdown file is reproduced from the embedded skeleton, and either the translated or source strings from the xliff file, depending on whether you want a translated or untranslated markdown file.
    * updateXliff: to update an existing xliff file with changes from a markdown file, ensuring that IDs of existing translatable strings are kept in tact. This command extracts the skeleton from the xliff file, makes a diff of the old and new markdown files, then applies this diff to the skeleton file I.e. removes skeleton lines that were removed from the markdown file, and adds skeleton lines (with new IDs) for lines that are newly added  to the markdown file. All existing lines stay as is keeping their existing translation IDs. Finally a new xliff file is generated from the up to date markdown file and skeleton, resulting in an xliff file that contains all translatable strings from the new markdown file, but reusing translation IDs for existing strings. 
    * translateXliff: given an xliff file, and a pretranslated markdown file that matches the skeleton, a new xliff file is produced containing translations for all strings.
    * pretranslateAllPossibleLangs: this walks the NVDA user_docs directory, and for each language, pretranslates the English xliff file using the existing pretranslated markdown file from the old translation system (if it matches the skeleton exactly) producing a translated xliff file that can be uploaded to Crowdin to bring an existing translation up to where it was in the old  system.
    * runTests: runs some tests using 2024.2 and 2024.3 English and French user guides to ensure files are translated and generated as expected.
* Added a generated xliff file for the current English user guide markdown file. Note that this has been uploaded to Crowdin for translation. 
* Added a GitHub action that runs on the beta branch if English userGuide.md changes. The action regenerates the original markdown file from the current English user guide xliff, then updates the xliff file based on the changes from the original markdown file to the current markdown file. This xliff file is then uploaded to Crowdin, and also committed and pushed to beta.
 
### Testing strategy:
* Uploaded the English user guide xliff to Crowdin.
* pretranslated a French xliff file from the English user guide xliff and a pretranslated French markdown file. 
* Uploaded the pretranslated French  xliff file to Crowdin for French.
* Downloaded the latest French xliff file from Crwdin, ensuring all strings were translated.
* Added / removed several lines from the English user guide markdown file and committed changes.
* Ensured the GitHub action ran with no errors.
* Downloded the French xliff from Crowdin and confirmed that the new added strings were available for translation, the removed strings were no longer available, and that all existing strings were still translated appropriately.

### Known issues with pull request:
None known.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Introduced an automated workflow for updating user documentation based on changes in markdown files, enhancing documentation consistency.
  
- **Bug Fixes**
  - Improved formatting and readability of various functions and classes across the codebase without altering functionality.
  
- **Documentation**
  - Enhanced clarity in comments and docstrings throughout the code, contributing to better maintainability. 

- **Chores**
  - General code style improvements, including consistent spacing and formatting adjustments to adhere to coding standards.

### CODE REVIEW

**Comment by michaelDCurran** [comment]
@seanbudd  It seems that precommit is reformatting all files in the repo which makes this pr impossible to review. I guess I'll need to wait for all the reformatting to reach beta and then remove the final precommit here before I can make it ready for review?

**Comment by CyrilleB79** [comment]
Before merging this PR, could you send a message to the translators mailing list and give a deadline for them to provide up-to-date translations?
You should explain that it's the last chance for them to reach the identical structure state, allowing to benefit from automatic translation in the new file.
Thanks.

**Comment by dpy013** [comment]
> Before merging this PR, could you send a message to the translators mailing list and give a deadline for them to provide up-to-date translations? You should explain that it's the last chance for them to reach the identical structure state, allowing to benefit from automatic translation in the new file. Thanks.
Suggest updating as well:
#15014 
Update on the plan

**Comment by michaelDCurran** [comment]
@CyrilleB79  and @dpy013   I'll definitely be providing a clear and detailed update to the translators list and the discussion once this PR is finally finalized approved and merged. But until then, things could change so I do not want to say anything prematurely. Please be assured though that we'll give translators plenty of time. Plus, the system is such that we can move the languages to Crowdin that are ready -- currently about 17. And when more become complete, we can move those as well.

File: user_docs/markdownTranslate.py [hunk 1]
```diff
@@ -0,0 +1,734 @@
+# A part of NonVisual Desktop Access (NVDA)
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:1]: I still need to review this file further

File: user_docs/markdownTranslate.py [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+	pretranslateLangsParser.add_argument(
+		"-b",
+		"--md-base-name",
+		dest="mdBaseName",
+		type=str,
+		required=True,
+		help="The base name of the markdown files to pretranslate",
+	)
+	args = mainParser.parse_args()
+	if args.command == "generateXliff":
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:712]: this could probably use match/case syntax

File: user_docs/markdownTranslate.py [hunk 3]
```diff
@@ -0,0 +1,734 @@
...
+				raise ValueError(f"Files do not match at line {lineNo}: {line1=} {line2=}")
+		print("Files match")
+
+
+def markdownTranslateCommand(command: str, *args):
+	print(f"Running markdownTranslate command: {command} {' '.join(args)}")
+	subprocess.run(["python", __file__, command, *args], check=True)
+
+
+def runTests(testDir: str):
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:483]: could these tests be moved to our standard unit tests?

File: user_docs/markdownTranslate.py [hunk 4]
```diff
@@ -0,0 +1,670 @@
...
+re_kcTitle = re.compile(r"^(<!--\s+KC:title:\s*)(.+?)(\s*-->)$")
+re_kcSettingsSection = re.compile(r"^(<!--\s+KC:settingsSection:\s*)(.+?)(\s*-->)$")
+re_comment = re.compile(r"^<!--.+-->$")
+re_heading = re.compile(r"^(#+\s+)(.+?)((?:\s+\{#.+\})?)$")
+re_bullet = re.compile(r"^(\s*\*\s+)(.+)$")
+re_number = re.compile(r"^(\s*[0-9]+\.\s+)(.+)$")
+re_hiddenHeaderRow = re.compile(r"^\|\s*\.\s*\{\.hideHeaderRow\}\s*(\|\s*\.\s*)*\|$")
+re_postTableHeaderLine = re.compile(r"^(\|\s*-+\s*)+\|$")
+re_tableRow = re.compile(r"^(\|)(.+)(\|)$")
+re_translationID = re.compile(r"^(.*)\$\(ID:([0-9a-f-]+)\)(.*)$")
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:31]: I think these should be grouped as an enum

```py
class Re(re.Pattern, Enum):
   KC_TITLE  = re.compile(r"^(<!--\s+KC:title:\s*)(.+?)(\s*-->)$")

File: user_docs/markdownTranslate.py [hunk 5]
```diff
@@ -0,0 +1,670 @@
...
+	result = subprocess.run(
+		["git", "rev-parse", "--show-toplevel"], capture_output=True, text=True, check=True
+	)
+	gitDir = result.stdout.strip()
+	if not os.path.isdir(gitDir):
+		raise ValueError(f"Invalid git directory: '{gitDir}'")
+	return gitDir
+
+
+def getRawGithubURLForPath(filePath: str):
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:75]:
```suggestion
def getRawGithubURLForPath(filePath: str) -> str:
```

File: user_docs/markdownTranslate.py [hunk 6]
```diff
@@ -0,0 +1,670 @@
...
+
+
+@dataclass
+class Result_generateMarkdown:
+	numTotalLines = 0
+	numTranslatableStrings = 0
+	numTranslatedStrings = 0
+
+
+def generateMarkdown(xliffPath: str, outputPath: str, translated: bool = True):
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:394]:
```suggestion
def generateMarkdown(xliffPath: str, outputPath: str, translated: bool = True) -> Result_generateMarkdown:
```

File: user_docs/markdownTranslate.py [hunk 7]
```diff
@@ -0,0 +1,670 @@
...
+			elif mdDiffLine.startswith("-"):
+				res.numRemovedLines += 1
+				origSkelLine = next(origSkelLines)
+				if re_translationID.match(origSkelLine):
+					res.numRemovedTranslationPlaceholders += 1
+			else:
+				raise ValueError(f"Unexpected diff line: {mdDiffLine}")
+		print(
+			f"Updated skeleton file with {res.numAddedLines} added lines ({res.numAddedTranslationPlaceholders} translation placeholders), {res.numRemovedLines} removed lines ({res.numRemovedTranslationPlaceholders} translation placeholders), and {res.numUnchangedLines} unchanged lines ({res.numUnchangedTranslationPlaceholders} translation placeholders)"
+		)
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:205]:
```suggestion
		print(
			f"Updated skeleton file with {res.numAddedLines} added lines "
			f"({res.numAddedTranslationPlaceholders} translation placeholders), "
			f"{res.numRemovedLines} removed lines ({res.numRemovedTranslationPlaceholders} translation placeholders), "
			f"and {res.numUnchangedLines} unchanged lines ({res.numUnchangedTranslationPlaceholders} translation placeholders)"
		)
```

File: user_docs/markdownTranslate.py (hunk 4 above)
Comment (seanbudd) [reply user_docs/markdownTranslate.py:31]: @michaelDCurran this is the only remaining feedback from me

File: user_docs/markdownTranslate.py [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+			generateMarkdown(
+				xliffPath=langXliffPath,
+				outputPath=rebuiltLangMdPath,
+			)
+		except Exception as e:
+			print(f"Failed to rebuild {langDir} markdown: {e}")
+			os.remove(langXliffPath)
+			continue
+		try:
+			ensureFilesMatch(rebuiltLangMdPath, langPretranslatedMdPath, allowBadAnchors=True)
```
Comment (seanbudd) [inline user_docs/markdownTranslate.py:531]:
```suggestion
			ensureMarkdownFilesMatch(rebuiltLangMdPath, langPretranslatedMdPath, allowBadAnchors=True)
```

File: user_docs/markdownTranslate.py (hunk 4 above)
Comment (michaelDCurran) [reply user_docs/markdownTranslate.py:31]: A few issues when trying to do this:
* At runtime:
```
class MarkdownRegex(re.Pattern, enum.Enum):                                                                                                             
  File "C:\Users\mick\AppData\Local\Programs\Python\Python311-32\Lib\enum.py", line 562, in __new__                                                         
    enum_class = super().__new__(metacls, cls, bases, classdict, **kwds)                                                                                    
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                    
TypeError: type 're.Pattern' is not an acceptable base type 
```
* In VS code, Pylance says:
```
Error: Base class "Pattern[Unknown]" is marked final and cannot be subclassed at line 25 and character 21. generated by Pylance
```
* If I don't inherit from re.Pattern however, then I of course get errors in VS Code like:
```
Error: Cannot access attribute "match" for class "Literal[MarkdownRegex.HIDDEN_HEADER_ROW]"   Attribute "match" is unknown at line 93 and character 38. generated by Pylance
```
So based on this, for now perhaps we should leave them as stand-alone constants?

**Review by seanbudd — APPROVED** [review]

File: user_docs/markdownTranslate.py (hunk 4 above)
Comment (seanbudd) [reply user_docs/markdownTranslate.py:31]: Makes sens

**Comment by michaelDCurran** [comment]
@coderabbitai resolve

**Pull request merged**